Bottom: 754be25be7f1808e265bf2e8a65d190fdeaa75da
Top:    9305abf9e4b6c4a47bc9d9dc5571b04a3add44c6
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-06-05 12:12:36 +0800

vhost: allow to disable device IOTLB

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index e3d7ea1..b18c580 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -1220,6 +1220,7 @@ static long vhost_net_reset_owner(struct vhost_net *n)
 static int vhost_net_set_features(struct vhost_net *n, u64 features)
 {
 	size_t vhost_hlen, sock_hlen, hdr_len;
+	int device_iotlb = features & (1ULL << VIRTIO_F_IOMMU_PLATFORM);
 	int i;
 
 	hdr_len = (features & ((1ULL << VIRTIO_NET_F_MRG_RXBUF) |
@@ -1240,10 +1241,8 @@ static int vhost_net_set_features(struct vhost_net *n, u64 features)
 	    !vhost_log_access_ok(&n->dev))
 		goto out_unlock;
 
-	if ((features & (1ULL << VIRTIO_F_IOMMU_PLATFORM))) {
-		if (vhost_init_device_iotlb(&n->dev, true))
-			goto out_unlock;
-	}
+	if (vhost_init_device_iotlb(&n->dev, device_iotlb))
+		goto out_unlock;
 
 	for (i = 0; i < VHOST_NET_VQ_MAX; ++i) {
 		mutex_lock(&n->vqs[i].vq.mutex);
diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 042030e..03c0572 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -1566,12 +1566,14 @@ EXPORT_SYMBOL_GPL(vhost_vring_ioctl);
 
 int vhost_init_device_iotlb(struct vhost_dev *d, bool enabled)
 {
-	struct vhost_umem *niotlb, *oiotlb;
+	struct vhost_umem *niotlb = NULL, *oiotlb;
 	int i;
 
-	niotlb = vhost_umem_alloc();
-	if (!niotlb)
-		return -ENOMEM;
+	if (enabled) {
+		niotlb = vhost_umem_alloc();
+		if (!niotlb)
+			return -ENOMEM;
+	}
 
 	oiotlb = d->iotlb;
 	d->iotlb = niotlb;
